File: src/MsSql/waitFor.ts
import { setTimeout as pause } from "node:timers/promises";
type MaybePromise<T> = T | Promise<T>;

interface WaitForOpts {
	signal?: AbortSignal;
	preDelay?: number;
	minDelay?: number;
	maxDelay?: number;
	watchDogMs?: number;
	rate?: number;
}

/** truncated exponential backoff polling, until returns truthy */
export async function waitFor(
	exp: () => MaybePromise<unknown>,
	{ signal, watchDogMs = 60_000, minDelay = 200, maxDelay = 5_000, rate = 1.5, preDelay }: WaitForOpts = {}
): Promise<void> {
	if (!Number.isFinite(rate) || rate < 1.0 || rate > 10) {
		throw new RangeError(`Rate for exp backof must be in range 1 <= RATE <= 10, got ${rate}`);
	}
	if (!Number.isInteger(minDelay) || minDelay < 1) {
		throw new TypeError(`mindelay value must be a positive integer, got ${minDelay}`);
	}
	if (!Number.isInteger(maxDelay) || maxDelay < minDelay) {
		throw new TypeError(
			"maxDelay value must be a positive integer bigger than mindelay, " + `got ${maxDelay} (mindelay = ${minDelay})`
		);
	}
	if (!Number.isInteger(watchDogMs) || watchDogMs < maxDelay) {
		throw new TypeError(
			"watchDogMs value must be a positive integer bigger than maxDelay, " +
				`got ${watchDogMs} (maxDelay = ${maxDelay})`
		);
	}

	let currentDelay = minDelay;
	let nRuns = 0;
	if (preDelay) {
		await pause(preDelay, undefined, { signal });
	}
	await withTimeout(async () => {
		for (;;) {
			++nRuns;
			if (signal?.aborted) {
				throw signal.reason ?? new Error("waitFor was aborted without a reason");
			}
			// add ExecutionMax, and abort execution after it's passed, in case of long
			// hanging promises
			const result = await exp();
			if (result) break;
			await pause(currentDelay, undefined, { signal });
			currentDelay = Math.min(currentDelay * rate, maxDelay);
		}
	}, watchDogMs);
}

function withTimeout(cb: () => Promise<unknown>, timeout = 60_000): Promise<void> {
	return new Promise<void>((res, rej) => {
		const to = setTimeout(() => rej(new TimeoutError()), timeout);
		cb()
			.then(() => res())
			.catch(rej)
			.finally(() => {
				clearTimeout(to);
			});
	});
}

class TimeoutError extends Error {
	override name = "TimeoutError";
	constructor(msg?: string, opts?: ErrorOptions) {
		super(msg ?? "TimeoutError", opts);
	}
}

File: mssql/waitFor.go
package mssql

import (
	"context"
	"fmt"
	"math"
	"time"
)

type WaitForOpts struct {
	InitialDelay int
	MaxDelay     int
	Rate         float64
	Immediate    bool
}

func waitFor(ctx context.Context, exp func() error) error {
	return waitForWithOpts(ctx, exp, WaitForOpts{})
}

func waitForWithOpts(ctx context.Context, exp func() error, opts WaitForOpts) error {
	if opts.InitialDelay == 0 {
		opts.InitialDelay = 200
	}
	if opts.MaxDelay == 0 {
		opts.MaxDelay = 5000
	}
	if opts.Rate == 0 {
		opts.Rate = 1.5
	}

	if 1.0 > opts.Rate || opts.Rate > 10 {
		return fmt.Errorf("rate for exp backoff must be in range 1 <= RATE <= 10, got %v", opts.Rate)
	}
	if opts.InitialDelay < 1 {
		return fmt.Errorf("initial delay value must be a positive integer, got %d", opts.InitialDelay)
	}
	if opts.MaxDelay < opts.InitialDelay {
		return fmt.Errorf("max delay value must be a positive integer bigger than initialDelay, got %d (initialDelay = %d)", opts.MaxDelay, opts.InitialDelay)
	}

	currentDelay := time.Duration(opts.InitialDelay) * time.Millisecond
	if !opts.Immediate {
		time.Sleep(currentDelay)
		currentDelay = time.Duration(math.Min(float64(currentDelay)*opts.Rate, float64(opts.MaxDelay))) * time.Millisecond
	}

	for {
		select {
		case <-ctx.Done():
			return fmt.Errorf("operation timed out: %w", ctx.Err())
		default:
			if err := exp(); err == nil {
				return nil
			}
			time.Sleep(currentDelay)
			currentDelay = time.Duration(math.Min(float64(currentDelay)*opts.Rate, float64(opts.MaxDelay))) * time.Millisecond
		}
	}
}
